File: Micro-Services/pong_project/pong_project/static/js/pong/sound.js
import * as g from './global.js';

class SoundEvent {
	constructor(tick, sound) {
		this.tick = tick;
		this.sound = sound;
		this.played = false;
	}
}

export class SoundManager {
	constructor() {
		this.events = [];
		this.music = new Audio(window.ASSETS_URL + "glitchstairs.ogg");
		this.victory = new Audio(window.ASSETS_URL + "victory.wav");
		this.hit_sound = new Audio(window.ASSETS_URL + "hit.wav");
		this.score_sound = new Audio(window.ASSETS_URL + "explosion.wav");
	}

	create_sound_events(arr, len) {
		for (let i = 0; i < len; i++) {
			this.events.push(new SoundEvent(arr[i][0], arr[i][1]));
		}
	}

	process_sound_events(tick) {
		this.events = this.events.filter(event => event.played === false);

		for (let i = 0; i < this.events.length; i++) {
			if (tick >= this.events[i].tick) {

				switch (this.events[i].sound) {
					case g.SOUND_HIT:
						this.play_hit_sound();
						break;
					case g.SOUND_SCORE:
						this.play_score_sound();
						break;
					case g.SOUND_VICTORY:
						this.play_victory_sound();
						break;
				}
				this.events[i].played = true;
			}
		}
	}

	play_hit_sound() {
		this.hit_sound.play().catch(error => {
			console.warn("Audio play failed:", error);
		});
	}

	play_score_sound() {
		this.score_sound.play().catch(error => {
			console.warn("Audio play failed:", error);
		});
	}

	play_victory_sound() {
		this.victory.play().catch(error => {
			console.warn("Audio play failed:", error);
		});
	}

	play_music() {
		this.music.loop = true;
		this.music.play().catch(error => {
			console.warn("Audio play failed:", error);
		});
	}

	stop_music() {
		this.music.pause();
		this.music.currentTime = 0;
	}
}

File: Micro-Services/game_app/game_app/pong/sound.py
import game_app.pong.constants as g

class SoundEvent:
    def __init__(self, tick, sound) -> None:
        self.tick = tick
        self.sound = sound

class SoundManager:
    def __init__(self) -> None:
        self.sound_events = []

    def create_sound_events(self, tick, collision, score, victory):
        if collision:
            self.sound_events.append(SoundEvent(tick, g.SOUND_HIT))
        if score:
            self.sound_events.append(SoundEvent(tick, g.SOUND_SCORE))
        if victory:
            self.sound_events.append(SoundEvent(tick, g.SOUND_VICTORY))

        self.sound_events = [event for event in self.sound_events if tick - event.tick < 3]


    def serialize_sound_events(self):
        serialized_events = [(event.tick, event.sound) for event in self.sound_events]
        return serialized_events